File: java/src/main/java/org/endeavourhealth/library/model/dto/CodeGenDto.java
package org.endeavourhealth.library.model.dto;

import com.fasterxml.jackson.annotation.JsonIgnore;

import java.util.HashMap;
import java.util.Map;

public class CodeGenDto {
  private String name;
  private String extension;
  private String collectionWrapper;
  private Map<String, String> datatypeMap = new HashMap<>();
  private String template;
  private Boolean complexTypes = false;

  public CodeGenDto() {
  }

  public String getName() {
    return name;
  }

  public CodeGenDto setName(String name) {
    this.name = name;
    return this;
  }

  public String getExtension() {
    return extension;
  }

  public CodeGenDto setExtension(String extension) {
    this.extension = extension;
    return this;
  }

  public String getCollectionWrapper() {
    return collectionWrapper;
  }

  public CodeGenDto setCollectionWrapper(String collectionWrapper) {
    this.collectionWrapper = collectionWrapper;
    return this;
  }

  public Map<String, String> getDatatypeMap() {
    return datatypeMap;
  }

  public CodeGenDto setDatatypeMap(Map<String, String> datatypeMap) {
    this.datatypeMap = datatypeMap;
    return this;
  }

  @JsonIgnore
  public String getDataType(String datatype) {
    if (datatypeMap == null)
      return null;

    return datatypeMap.get(datatype);
  }

  public String getTemplate() {
    return template;
  }

  public CodeGenDto setTemplate(String template) {
    this.template = template;
    return this;
  }

  public Boolean getComplexTypes() {
    return complexTypes;
  }

  public CodeGenDto setComplexTypes(Boolean complexTypes) {
    this.complexTypes = complexTypes;
    return this;
  }

  @JsonIgnore
  public boolean hasCollectionWrapper() {
    return collectionWrapper != null && !collectionWrapper.isEmpty();
  }
}

File: src/interfaces/CodeGenDto.ts
import z from "zod";

// export interface CodeGenDto {
//   name?: string;
//   extension?: string;
//   collectionWrapper?: string;
//   datatypeMap?: { [index: string]: string };
//   template?: string;
//   complexTypes?: boolean;
// }

export const CodeGenDtoSchema = z.strictObject({
  name: z.string().optional(),
  extension: z.string().optional(),
  collectionWrapper: z.string().optional(),
  datatypeMap: z.map(z.string(), z.string()).default(new Map()),
  template: z.string().optional(),
  complexTypes: z.boolean().default(false)
});

export type CodeGenDto = z.output<typeof CodeGenDtoSchema>;
